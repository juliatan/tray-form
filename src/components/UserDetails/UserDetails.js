import React from 'react';
import { Formik, Field, Form, ErrorMessage } from 'formik';
import * as Yup from 'yup';
import classes from './UserDetails.module.scss';

const userDetails = (props) => {
  const UserDetailsSchema = Yup.object().shape({
    name: Yup.string()
      .max(50, 'Please use less than 50 characters')
      .required('Your name is required'),
    password: Yup.string()
      .min(10, 'Password needs to be a minimum of 10 characters')
      .max(50, 'Please use less than 50 characters')
      .matches(
        /^(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.{10,})/,
        'Password needs to contain at least one uppercase, one lowercase and one number',
      )
      .required('Your password is required'),
    email: Yup.string()
      .email("This doesn't look like a valid email address")
      .required('Your email is required'),
  });

  return (
    <div className={classes.UserDetails}>
      <Formik
        initialValues={{ email: '', password: '', name: '', role: '' }}
        validationSchema={UserDetailsSchema}
        onSubmit={(values, { setSubmitting }) => {
          props.saveUserDetails(values);
          setSubmitting(false);
        }}
      >
        {({ isSubmitting, isValid, dirty }) => (
          <Form>
            <label htmlFor="name">Name:</label>
            <Field name="name" type="text" placeholder="Full name" />
            <ErrorMessage name="name" />

            <label htmlFor="role">Role:</label>
            <Field name="role" type="text" placeholder="Role" />
            <ErrorMessage name="role" />

            <label htmlFor="email">Email:</label>
            <Field name="email" type="email" placeholder="Email address" />
            <ErrorMessage name="email" />

            <label htmlFor="password">Password:</label>
            <Field
              name="password"
              type="password"
              placeholder="Password (minimum of 10 characters)"
            />
            <ErrorMessage name="password" />

            <button
              type="submit"
              disabled={!(isValid && dirty) || isSubmitting}
            >
              Submit
            </button>
          </Form>
        )}
      </Formik>
    </div>
  );
};

export default userDetails;
