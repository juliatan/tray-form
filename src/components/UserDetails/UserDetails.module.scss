.UserDetails {
  form {
    display: flex;
    flex-direction: column;
  }

  input {
    margin: 8px 0;
    height: 40px;
    border: 1px solid black;
    border-radius: 4px;
  }
}
