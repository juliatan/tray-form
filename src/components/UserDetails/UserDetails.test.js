import React from 'react';
import { render, cleanup, fireEvent, waitFor } from '@testing-library/react';
import UserDetails from './UserDetails';

describe('UserDetails component', () => {
  let saveDetails;
  let component;

  beforeEach(() => {
    saveDetails = jest.fn();
    component = <UserDetails saveDetails={saveDetails} />;
  });

  afterEach(cleanup);

  it('should render', () => {
    const { asFragment } = render(component);
    expect(asFragment(component)).toMatchSnapshot();
  });

  it('should submit form with validated user details', async () => {
    const { container, getByTestId } = render(component);
    const nameInput = container.querySelector('input[name="name"]');
    const roleInput = container.querySelector('input[name="role"]');
    const emailInput = container.querySelector('input[name="email"]');
    const passwordInput = container.querySelector('input[name="password"]');

    const name = 'mockname';
    const role = 'mockrole';
    const email = 'test@email.com';
    const password = '1234asdfASDF';

    fireEvent.change(nameInput, {
      target: {
        value: name,
      },
    });

    fireEvent.change(roleInput, {
      target: {
        value: role,
      },
    });

    fireEvent.change(emailInput, {
      target: {
        value: email,
      },
    });

    fireEvent.change(passwordInput, {
      target: {
        value: password,
      },
    });

    fireEvent.click(getByTestId('userDetailsSubmit'));
    await waitFor(() =>
      expect(saveDetails).toHaveBeenCalledWith({
        name,
        role,
        email,
        password,
      }),
    );
  });

  it('should not submit form if no name', async () => {
    const { container, getByTestId } = render(component);
    const emailInput = container.querySelector('input[name="email"]');
    const passwordInput = container.querySelector('input[name="password"]');

    const email = 'test@email.com';
    const password = '1234asdfASDF';

    fireEvent.change(emailInput, {
      target: {
        value: email,
      },
    });

    fireEvent.change(passwordInput, {
      target: {
        value: password,
      },
    });

    fireEvent.click(getByTestId('userDetailsSubmit'));
    await waitFor(() => expect(saveDetails).not.toHaveBeenCalled());
  });

  it('should not submit form if invalid email', async () => {
    const { container, getByTestId } = render(component);
    const nameInput = container.querySelector('input[name="name"]');
    const emailInput = container.querySelector('input[name="email"]');
    const passwordInput = container.querySelector('input[name="password"]');

    const name = 'mockname';
    const email = 'notanemail';
    const password = '1234asdfASDF';

    fireEvent.change(nameInput, {
      target: {
        value: name,
      },
    });

    fireEvent.change(emailInput, {
      target: {
        value: email,
      },
    });

    fireEvent.change(passwordInput, {
      target: {
        value: password,
      },
    });

    fireEvent.click(getByTestId('userDetailsSubmit'));
    await waitFor(() => expect(saveDetails).not.toHaveBeenCalled());
  });

  it('should not submit form if invalid password', async () => {
    const { container, getByTestId } = render(component);
    const nameInput = container.querySelector('input[name="name"]');
    const emailInput = container.querySelector('input[name="email"]');
    const passwordInput = container.querySelector('input[name="password"]');

    const name = 'mockname';
    const email = 'test@email.com';
    const password = 'poorpassword';

    fireEvent.change(nameInput, {
      target: {
        value: name,
      },
    });

    fireEvent.change(emailInput, {
      target: {
        value: email,
      },
    });

    fireEvent.change(passwordInput, {
      target: {
        value: password,
      },
    });

    fireEvent.click(getByTestId('userDetailsSubmit'));
    await waitFor(() => expect(saveDetails).not.toHaveBeenCalled());
  });
});
