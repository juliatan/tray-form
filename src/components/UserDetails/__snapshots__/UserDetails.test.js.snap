// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserDetails component should render 1`] = `
<DocumentFragment>
  <div
    class="UserDetails"
  >
    <p
      class="Legend"
    >
      * is a required field
    </p>
    <form
      action="#"
    >
      <div
        class="LabelContainer"
      >
        <label
          for="name"
        >
          Name: 
          <span
            class="Required"
          >
            *
          </span>
        </label>
      </div>
      <input
        name="name"
        placeholder="Jane Doe"
        type="text"
        value=""
      />
      <div
        class="LabelContainer"
      >
        <label
          for="role"
        >
          Role:
        </label>
      </div>
      <input
        name="role"
        placeholder="Engineer"
        type="text"
        value=""
      />
      <div
        class="LabelContainer"
      >
        <label
          for="email"
        >
          Email: 
          <span
            class="Required"
          >
            *
          </span>
        </label>
      </div>
      <input
        name="email"
        placeholder="name@company.com"
        type="email"
        value=""
      />
      <div
        class="LabelContainer"
      >
        <label
          for="password"
        >
          Password: 
          <span
            class="Required"
          >
            *
          </span>
        </label>
      </div>
      <input
        name="password"
        placeholder="Min. 10 chars"
        type="password"
        value=""
      />
      <button
        class="Button"
        data-testid="userDetailsSubmit"
        disabled=""
        type="submit"
      >
        Submit
      </button>
    </form>
  </div>
</DocumentFragment>
`;
